// lucid-extension-sdk/core/spreadsheetintegration/lucidspreadsheetintegration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LucidSpreadsheetIntegration = exports.CustomDetailsChooser = exports.ListChooser = exports.isImportedResults = exports.lucidSpreadsheetIntegrationFailureTypeValidator = exports.LucidSpreadsheetIntegrationFailureType = void 0;
const checks_1 = require("../checks");
const serializedupstreamconfig_1 = require("../data/datasource/serializedupstreamconfig");
const serializedschema_1 = require("../data/serializedfield/serializedschema");
const jsonserializable_1 = require("../jsonserializable");
const validators_1 = require("../validators/validators");
/** @ignore until spreadsheet integration is ready for launch (CHART-51946) */
var LucidSpreadsheetIntegrationFailureType;
(function (LucidSpreadsheetIntegrationFailureType) {
    LucidSpreadsheetIntegrationFailureType["AuthorizationFailure"] = "AuthorizationFailure";
    LucidSpreadsheetIntegrationFailureType["GenericFailure"] = "GenericFailure";
    LucidSpreadsheetIntegrationFailureType["UserCancelled"] = "UserCanceled";
    LucidSpreadsheetIntegrationFailureType["PopupsNotAllowed"] = "PopupsNotAllowed";
})(LucidSpreadsheetIntegrationFailureType || (exports.LucidSpreadsheetIntegrationFailureType = LucidSpreadsheetIntegrationFailureType = {}));
/** @ignore until spreadsheet integration is ready for launch (CHART-51946) */
exports.lucidSpreadsheetIntegrationFailureTypeValidator = (0, validators_1.stringEnumValidator)(LucidSpreadsheetIntegrationFailureType);
/** @ignore until spreadsheet integration is ready for launch (CHART-51946) */
exports.isImportedResults = (0, validators_1.strictObjectValidator)({
    'upstreamConfig': serializedupstreamconfig_1.isSerializedUpstreamConfig,
    'collections': (0, validators_1.arrayValidator)((0, validators_1.strictObjectValidator)({
        'collectionId': checks_1.isString,
        'headerRow': checks_1.isNumber,
        'schema': serializedschema_1.isSerializedSchema,
        'upstreamConfig': jsonserializable_1.isJsonSerializable,
    })),
});
/** @ignore until spreadsheet integration is ready for launch (CHART-51946) */
class ListChooser {
    constructor(getList) {
        this.getList = getList;
    }
}
exports.ListChooser = ListChooser;
/** @ignore until spreadsheet integration is ready for launch (CHART-51946) */
class CustomDetailsChooser {
    constructor(getConfig) {
        this.getConfig = getConfig;
    }
}
exports.CustomDetailsChooser = CustomDetailsChooser;
/** @ignore until spreadsheet integration is ready for launch (CHART-51946) */
class LucidSpreadsheetIntegration {
    constructor() {
        this.addCardsIntegration = true;
    }
}
exports.LucidSpreadsheetIntegration = LucidSpreadsheetIntegration;
