// lucid-extension-sdk/core/cardintegration/cardfielddisplaysettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LucidCardFields = exports.StatusValues = exports.deserializeLucidCardFieldDisplaySettings = exports.serializeLucidCardFieldDisplaySettings = exports.isSerializedLucidCardFieldSettings = exports.isOnClickHandlerKey = exports.OnClickHandlerKeys = exports.isFieldDisplayType = exports.FieldDisplayType = void 0;
const checks_1 = require("../checks");
const badgeposition_1 = require("../properties/datagraphic/badgeposition");
const validators_1 = require("../validators/validators");
/**
 * When configuring a field on a LucidCardBlock to be displayed as a data graphic, one of these values specifies
 * the function to convert the field value to the data graphic settings. The definition of these functions is in
 * LucidCardsLibrary.registerStencilDisplayTypes.
 */
var FieldDisplayType;
(function (FieldDisplayType) {
    /**
     * Text badge with the full text of the field.
     * Black text on a light gray background in the upper-right of the card.
     */
    FieldDisplayType["BasicTextBadge"] = "BasicTextBadge";
    /**
     * Given a name (usually of a user), puts a text badge with their initials (first letter of each name)
     * at the bottom-left of the card, black text on a light gray background.
     */
    FieldDisplayType["InitializedString"] = "InitializedString";
    /**
     * Given a number, display the number in the bottom-left of the card. If the number is higher than 999,
     * display 999.
     */
    FieldDisplayType["StandardEstimation"] = "StandardEstimation";
    /**
     * Given a URL, display a small image cropped into a circle in the lower-left of the card.
     */
    FieldDisplayType["ImageBadge"] = "ImageBadge";
    /**
     * Given a URL, display a small image cropped into a square in the lower-left of the card.
     */
    FieldDisplayType["SquareImageBadge"] = "SquareImageBadge";
    /**
     * Given a date, display a small calendar icon alongside a very short version of the date
     * as a string, e.g. "Sep 9"
     */
    FieldDisplayType["DateBadge"] = "DateBadge";
    /** Given an object containing iconUrl (optional) and name (optional), determine whether
     * to display as an ImageBadge or InitializedString (or nothing, if neither is present).
     *
     * If FF LucidCardUIRefresh ON:
     * Given or string of name or an object containing iconUrl and name, display as
     * an as a full name or an ImageBadge with full name (if name is present).
     */
    FieldDisplayType["UserProfile"] = "UserProfile";
})(FieldDisplayType || (exports.FieldDisplayType = FieldDisplayType = {}));
exports.isFieldDisplayType = (0, validators_1.enumValidator)(FieldDisplayType);
/**
 * When configuring a field on a LucidCardBlock to be displayed as a data graphic, this click handler value
 * specifies what to do when the user clicks on the data graphic produced.
 */
var OnClickHandlerKeys;
(function (OnClickHandlerKeys) {
    /**
     * Display the UI panel to edit fields on the data associated with the card
     */
    OnClickHandlerKeys["BasicEditPanel"] = "BasicEditPanel";
    /**
     * Open a new browser window to a URL specified by linkFormula
     */
    OnClickHandlerKeys["OpenBrowserWindow"] = "OpenBrowserWindow";
    /**
     * Display a custom UI panel to edit non-standard fields on the data associated with the card
     */
    OnClickHandlerKeys["CustomEditAction"] = "CustomEditAction";
})(OnClickHandlerKeys || (exports.OnClickHandlerKeys = OnClickHandlerKeys = {}));
exports.isOnClickHandlerKey = (0, validators_1.enumValidator)(OnClickHandlerKeys);
const isImageBadgeSettings = (0, validators_1.objectValidator)({
    'height': checks_1.isNumber,
    'width': checks_1.isNumber,
});
const isTextBadgeSettings = (0, validators_1.objectValidator)({
    'bold': (0, validators_1.option)(checks_1.isBoolean),
    'italic': (0, validators_1.option)(checks_1.isBoolean),
    'underline': (0, validators_1.option)(checks_1.isBoolean),
    'rounding': (0, validators_1.option)(checks_1.isNumber),
    'minwidth': (0, validators_1.option)(checks_1.isNumber),
    'minheight': (0, validators_1.option)(checks_1.isNumber),
    'maxfontsize': (0, validators_1.option)(checks_1.isNumber),
    'maxtextwidth': (0, validators_1.option)(checks_1.isNumber),
    'iswithinpill': (0, validators_1.option)(checks_1.isBoolean),
    'usingcarduirefresh': (0, validators_1.option)(checks_1.isBoolean),
    'hasleadingicon': (0, validators_1.option)(checks_1.isBoolean),
});
const isStencilConfig = (0, validators_1.objectValidator)({
    'getterKey': exports.isFieldDisplayType,
    'fg': (0, validators_1.option)(checks_1.isString),
    'bg': (0, validators_1.option)(checks_1.isString),
    'f': (0, validators_1.option)(checks_1.isString),
    't': (0, validators_1.option)(checks_1.isString),
    'l': (0, validators_1.option)(checks_1.isString),
    'h': (0, validators_1.option)((0, validators_1.enumValidator)(badgeposition_1.HorizontalBadgePos)),
    'v': (0, validators_1.option)((0, validators_1.enumValidator)(badgeposition_1.VerticalBadgePos)),
    'onClickHandlerKey': (0, validators_1.option)(exports.isOnClickHandlerKey),
    'imageBadgeSettings': (0, validators_1.option)(isImageBadgeSettings),
    'textBadgeSettings': (0, validators_1.option)(isTextBadgeSettings),
});
exports.isSerializedLucidCardFieldSettings = (0, validators_1.objectValidator)({
    'StencilConfig': (0, validators_1.option)(isStencilConfig),
});
/** @ignore */
function serializeLucidCardFieldDisplaySettings(settings) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
    return {
        'StencilConfig': settings.stencilConfig && {
            'getterKey': settings.stencilConfig.displayType,
            'f': settings.stencilConfig.valueFormula,
            't': settings.stencilConfig.tooltipFormula,
            'l': settings.stencilConfig.linkFormula,
            'fg': settings.stencilConfig.foregroundColor,
            'bg': settings.stencilConfig.backgroundColor,
            'h': settings.stencilConfig.horizontalPosition,
            'v': settings.stencilConfig.verticalPosition,
            'onClickHandlerKey': settings.stencilConfig.onClickHandlerKey,
            'imageBadgeSettings': settings.stencilConfig.imageBadgeSettings && {
                'height': settings.stencilConfig.imageBadgeSettings.height,
                'width': settings.stencilConfig.imageBadgeSettings.width,
            },
            'textBadgeSettings': {
                'rounding': (_a = settings.stencilConfig.textBadgeSettings) === null || _a === void 0 ? void 0 : _a.rounding,
                'bold': (_b = settings.stencilConfig.textBadgeSettings) === null || _b === void 0 ? void 0 : _b.bold,
                'italic': (_c = settings.stencilConfig.textBadgeSettings) === null || _c === void 0 ? void 0 : _c.italic,
                'underline': (_d = settings.stencilConfig.textBadgeSettings) === null || _d === void 0 ? void 0 : _d.underline,
                'minwidth': (_e = settings.stencilConfig.textBadgeSettings) === null || _e === void 0 ? void 0 : _e.minwidth,
                'minheight': (_f = settings.stencilConfig.textBadgeSettings) === null || _f === void 0 ? void 0 : _f.minheight,
                'maxfontsize': (_g = settings.stencilConfig.textBadgeSettings) === null || _g === void 0 ? void 0 : _g.maxfontsize,
                'maxtextwidth': (_h = settings.stencilConfig.textBadgeSettings) === null || _h === void 0 ? void 0 : _h.maxtextwidth,
                'iswithinpill': (_j = settings.stencilConfig.textBadgeSettings) === null || _j === void 0 ? void 0 : _j.iswithinpill,
                'usingcarduirefresh': (_k = settings.stencilConfig.textBadgeSettings) === null || _k === void 0 ? void 0 : _k.usingcarduirefresh,
                'hasleadingicon': (_l = settings.stencilConfig.textBadgeSettings) === null || _l === void 0 ? void 0 : _l.hasleadingicon,
            },
        },
    };
}
exports.serializeLucidCardFieldDisplaySettings = serializeLucidCardFieldDisplaySettings;
/** @ignore */
function deserializeLucidCardFieldDisplaySettings(settings) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
    return {
        stencilConfig: settings['StencilConfig'] && {
            displayType: settings['StencilConfig']['getterKey'],
            valueFormula: settings['StencilConfig']['f'],
            tooltipFormula: settings['StencilConfig']['t'],
            linkFormula: settings['StencilConfig']['l'],
            foregroundColor: settings['StencilConfig']['fg'],
            backgroundColor: settings['StencilConfig']['bg'],
            horizontalPosition: settings['StencilConfig']['h'],
            verticalPosition: settings['StencilConfig']['v'],
            onClickHandlerKey: settings['StencilConfig']['onClickHandlerKey'],
            imageBadgeSettings: settings['StencilConfig']['imageBadgeSettings'] && {
                height: settings['StencilConfig']['imageBadgeSettings']['height'],
                width: settings['StencilConfig']['imageBadgeSettings']['width'],
            },
            textBadgeSettings: {
                rounding: (_a = settings['StencilConfig']['textBadgeSettings']) === null || _a === void 0 ? void 0 : _a['rounding'],
                bold: (_b = settings['StencilConfig']['textBadgeSettings']) === null || _b === void 0 ? void 0 : _b['bold'],
                italic: (_c = settings['StencilConfig']['textBadgeSettings']) === null || _c === void 0 ? void 0 : _c['italic'],
                underline: (_d = settings['StencilConfig']['textBadgeSettings']) === null || _d === void 0 ? void 0 : _d['underline'],
                minwidth: (_e = settings['StencilConfig']['textBadgeSettings']) === null || _e === void 0 ? void 0 : _e['minwidth'],
                minheight: (_f = settings['StencilConfig']['textBadgeSettings']) === null || _f === void 0 ? void 0 : _f['minheight'],
                maxfontsize: (_g = settings['StencilConfig']['textBadgeSettings']) === null || _g === void 0 ? void 0 : _g['maxfontsize'],
                maxtextwidth: (_h = settings['StencilConfig']['textBadgeSettings']) === null || _h === void 0 ? void 0 : _h['maxtextwidth'],
                iswithinpill: (_j = settings['StencilConfig']['textBadgeSettings']) === null || _j === void 0 ? void 0 : _j['iswithinpill'],
                usingcarduirefresh: (_k = settings['StencilConfig']['textBadgeSettings']) === null || _k === void 0 ? void 0 : _k['usingcarduirefresh'],
                hasleadingicon: (_l = settings['StencilConfig']['textBadgeSettings']) === null || _l === void 0 ? void 0 : _l['hasleadingicon'],
            },
        },
    };
}
exports.deserializeLucidCardFieldDisplaySettings = deserializeLucidCardFieldDisplaySettings;
/**
 * These are the six possible values for the status of a basic card block.
 * Must be kept in sync with cake/app/webroot/ts/libraries/lucidcards/lucidcarddefaultstatus.ts
 */
var StatusValues;
(function (StatusValues) {
    StatusValues["New"] = "New";
    StatusValues["Todo"] = "To Do";
    StatusValues["InProgress"] = "In Progress";
    StatusValues["Done"] = "Done";
    StatusValues["NotDoing"] = "Not Doing";
    StatusValues["Blocked"] = "Blocked";
})(StatusValues || (exports.StatusValues = StatusValues = {}));
var LucidCardFields;
(function (LucidCardFields) {
    LucidCardFields["Title"] = "Title";
    LucidCardFields["Description"] = "Description";
    LucidCardFields["Assignee"] = "Assignee";
    LucidCardFields["Estimate"] = "Estimate";
    LucidCardFields["Status"] = "Status";
    LucidCardFields["StartTime"] = "StartTime";
    LucidCardFields["EndTime"] = "EndTime";
})(LucidCardFields || (exports.LucidCardFields = LucidCardFields = {}));
