// lucid-package/src/interactibilityutil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.multiSelect = exports.confirmPrompt = exports.inputPrompt = exports.selectPrompt = void 0;
const checkbox_1 = require("@inquirer/checkbox");
const confirm_1 = require("@inquirer/confirm");
const input_1 = require("@inquirer/input");
const select_1 = require("@inquirer/select");
async function selectPrompt(message, options) {
    return await (0, select_1.default)({
        message: message,
        choices: options.map((option, index) => ({ name: option, value: index })),
    });
}
exports.selectPrompt = selectPrompt;
async function inputPrompt(message) {
    return await (0, input_1.default)({ message: message });
}
exports.inputPrompt = inputPrompt;
async function confirmPrompt(message) {
    return await (0, confirm_1.default)({ message: message });
}
exports.confirmPrompt = confirmPrompt;
async function multiSelect(message, options) {
    return await (0, checkbox_1.default)({
        message: message,
        choices: options.map((option, index) => ({ name: option, value: index })),
    });
}
exports.multiSelect = multiSelect;
