// lucid-extension-sdk/sdk/ui/menu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Menu = exports.MenuLocation = exports.MenuType = void 0;
var MenuType;
(function (MenuType) {
    MenuType[MenuType["Main"] = 1] = "Main";
    /** The context menu that appears when the user right-clicks the canvas. */
    MenuType[MenuType["Context"] = 2] = "Context";
})(MenuType = exports.MenuType || (exports.MenuType = {}));
/**
 * Semantic locations to place a new menu item.
 */
var MenuLocation;
(function (MenuLocation) {
    /** Default location: a new top-level menu entry for this extension, or the end of the context menu */
    MenuLocation[MenuLocation["Extension"] = 1] = "Extension";
    MenuLocation[MenuLocation["Edit"] = 2] = "Edit";
    MenuLocation[MenuLocation["View"] = 3] = "View";
    MenuLocation[MenuLocation["Share"] = 4] = "Share";
    MenuLocation[MenuLocation["Export"] = 5] = "Export";
    MenuLocation[MenuLocation["Import"] = 6] = "Import";
})(MenuLocation = exports.MenuLocation || (exports.MenuLocation = {}));
class Menu {
    constructor(client) {
        this.client = client;
    }
    /**
     * Add a new menu item to execute custom code. The action must be registered with
     * [EditorClient.registerAction](/extension-sdk/#classes_editorclient-EditorClient_registeraction)
     * prior to using it in the menu.
     *
     * @param item The definition of the new menu item
     */
    addMenuItem(item) {
        if (item.action && !this.client.actionExists(item.action)) {
            throw new Error('Unregistered action: ' + item.action);
        }
        if (item.visibleAction && !this.client.actionExists(item.visibleAction)) {
            throw new Error('Unregistered action: ' + item.visibleAction);
        }
        if (item.disabledAction && !this.client.actionExists(item.disabledAction)) {
            throw new Error('Unregistered action: ' + item.disabledAction);
        }
        this.client.sendCommand("ami" /* CommandName.AddMenuItem */, {
            'l': item.label,
            'a': item.action,
            'v': item.visibleAction,
            'd': item.disabledAction,
            't': item.menuType,
            'loc': item.location,
            'f': item.file
                ? {
                    'a': item.file.action,
                    'ac': item.file.accept,
                    's': item.file.singleFileOnly,
                    'b': item.file.binary,
                }
                : undefined,
        });
    }
}
exports.Menu = Menu;
