// lucid-extension-sdk/sdk/document/blockclasses/linkunfurlblockproxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExperimentalLinkUnfurlBlockProxy = exports.LinkUnfurlBlockProxy = void 0;
const iframeutils_1 = require("../../core/iframe/iframeutils");
const unfurliframe_1 = require("../../core/unfurl/unfurliframe");
const blockproxy_1 = require("../blockproxy");
/**
 * This is the kind of block that appears when a link is pasted onto the canvas of a Lucid editor.
 */
class LinkUnfurlBlockProxy extends blockproxy_1.BlockProxy {
    /**
     * Returns the title on the block, which is the main text shown on the block.
     */
    getTitle() {
        return this.textAreas.get('t_LinkUnfurlTitle');
    }
    /**
     * Sets the title on the block, which is the main text shown on the block.
     */
    setTitle(title) {
        this.textAreas.set('t_LinkUnfurlTitle', title, { force: true });
    }
    /**
     * Returns the name of the service the link belongs to, such as "Lucid" or "Google", as displayed below the title on
     * the block.
     */
    getProviderName() {
        return this.textAreas.get('t_LinkUnfurlProviderName');
    }
    /**
     * Sets the URL to be loaded in an iframe when the user clicks the "Expand" action on the block.
     */
    setIframe(unfurlIframe) {
        const iframeAttributes = (0, unfurliframe_1.generateIframeAttributes)(unfurlIframe);
        this.properties.set('LinkUnfurlIframeHtml', (0, iframeutils_1.generateIFrameHTML)(iframeAttributes));
        this.properties.set('LinkUnfurlIframeHeight', iframeAttributes.height);
        this.properties.set('LinkUnfurlIframeWidth', iframeAttributes.width);
    }
    /**
     * Sets the main thumbnail on the block
     */
    setPreviewThumbnailUrl(thumbnailUrl) {
        this.properties.set('LinkUnfurlThumbnailUrl', thumbnailUrl);
    }
    /**
     * Sets additional thumbnails on the block
     */
    setThumbnailUrls(thumbnails) {
        this.properties.set('LinkUnfurlThumbnailUrls', thumbnails);
    }
    /**
     * @ignore
     * Initiates the PDF upload for this block. The result includes an uploadUrl which you can use to upload the PDF.
     * The PDF uploaded at that url will be assocaite with this block (and any copy of it) in the editor until it is refreshed.
     */
    async experimentalStartPDFUpload() {
        const result = await this.client.sendCommand("pdf" /* CommandName.StartPDFUploadRequest */, this.id);
        return result;
    }
}
exports.LinkUnfurlBlockProxy = LinkUnfurlBlockProxy;
LinkUnfurlBlockProxy.classNameRegex = /^LinkUnfurlBlock$/;
/**
 * @ignore
 * @deprecated Use LinkUnfurlBlockProxy instead.
 */
exports.ExperimentalLinkUnfurlBlockProxy = LinkUnfurlBlockProxy;
