// lucid-extension-sdk/dataconnector/actions/serializedpatchtypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.patchValidator = exports.patchSchemaValidator = exports.thirdPartyColumnPatchValidator = exports.thirdPartyColumnValidator = exports.patchItemValidator = void 0;
const checks_1 = require("../../core/checks");
const fieldtypedefinition_1 = require("../../core/data/fieldtypedefinition/fieldtypedefinition");
const validators_1 = require("../../core/validators/validators");
exports.patchItemValidator = (0, validators_1.objectValidator)({
    itemsAdded: (0, validators_1.objectOfValidator)(checks_1.isObjectUnsafe),
    itemsChanged: (0, validators_1.objectOfValidator)(checks_1.isObjectUnsafe),
    itemsDeleted: (0, validators_1.arrayValidator)(checks_1.isString),
    itemOrderChanged: (0, validators_1.nullableOption)((0, validators_1.arrayValidator)((0, validators_1.tupleValidator)(checks_1.isString, (0, validators_1.nullable)(checks_1.isString)))),
    syncSourceId: checks_1.isString,
    syncCollectionId: checks_1.isString,
    adapterType: (0, checks_1.isLiteral)('EXTENSION_API'),
    syncSchema: checks_1.isUnknown,
});
exports.thirdPartyColumnValidator = (0, validators_1.objectValidator)({
    name: checks_1.isString,
    fieldType: fieldtypedefinition_1.isSerializedFieldTypeDefinition,
});
exports.thirdPartyColumnPatchValidator = (0, validators_1.objectValidator)({
    name: (0, validators_1.nullableOption)(checks_1.isString),
    fieldType: (0, validators_1.nullableOption)(fieldtypedefinition_1.isSerializedFieldTypeDefinition),
});
exports.patchSchemaValidator = (0, validators_1.objectValidator)({
    columnsAdded: (0, validators_1.arrayValidator)(exports.thirdPartyColumnValidator),
    columnsChanged: (0, validators_1.mapValidator)(exports.thirdPartyColumnPatchValidator),
    columnsDeleted: (0, validators_1.arrayValidator)(checks_1.isString),
    columnOrdering: (0, validators_1.nullableOption)((0, validators_1.arrayValidator)((0, validators_1.tupleValidator)(checks_1.isString, (0, validators_1.nullable)(checks_1.isString)))),
    syncSourceId: checks_1.isString,
    syncCollectionId: checks_1.isString,
    adapterType: (0, checks_1.isLiteral)('EXTENSION_API'),
});
exports.patchValidator = (0, validators_1.either)(exports.patchItemValidator, exports.patchSchemaValidator);
